src/ray/rpc/node_manager_server.h

```diff
@@ -0,0 +1,62 @@
+#ifndef RAY_RPC_NODE_MANAGER_SERVER_H
+#define RAY_RPC_NODE_MANAGER_SERVER_H
+
+#include "ray/rpc/grpc_server.h"
+#include "ray/rpc/server_call.h"
+
+#include "src/ray/protobuf/node_manager.grpc.pb.h"
+#include "src/ray/protobuf/node_manager.pb.h"
+
+namespace ray {
+
+/// Interface of the `NodeManagerService`, see `src/ray/protobuf/node_manager.proto`.
+class NodeManagerServiceHandler {
+ public:
+  /// Handle a `ForwardTask` request.
+  /// The implementation can handle this request asynchronously. When hanling is done, the
+  /// `done_callback` should be called.
+  ///
+  /// \param[in] request The request message.
+  /// \param[out] reply The reply message.
+  /// \param[in] done_callback The callback to be called when the request is done.
+  virtual void HandleForwardTask(const ForwardTaskRequest &request,
+                                 ForwardTaskReply *reply,
+                                 RequestDoneCallback done_callback) = 0;
+};
+
+/// The `GrpcServer` for `NodeManagerService`.
+class NodeManagerServer : public GrpcServer {
+ public:
+  /// Construct a `NodeManagerServer`.
+  ///
+  /// \param[in] port See `GrpcServer`.
+  /// \param[in] handler The service handler that actually handle the requests.
+  NodeManagerServer(const uint32_t port, NodeManagerServiceHandler &service_handler)
+      : GrpcServer("NodeManager", port), service_handler_(service_handler){};
+
+  void RegisterServices(::grpc::ServerBuilder &builder) override {
+    /// Register `NodeManagerService`.
+    builder.RegisterService(&service_);
+  }
+
+  void InitServerCallFactories(
+      std::vector<std::unique_ptr<ServerCallFactory>> *server_call_factories) override {
+    // Initialize the factory for `ForwardTask` requests.
+    std::unique_ptr<ServerCallFactory> forward_task_call_factory(
+        new ServerCallFactoryImpl<NodeManagerService, NodeManagerServiceHandler,
+                                  ForwardTaskRequest, ForwardTaskReply>(
+            service_, &NodeManagerService::AsyncService::RequestForwardTask,
```

Comment:
a separate factory is required for each of the methods?

Reply:
Yeah, each factory object is responsible for one kind of RPC method

Reply:
And that's needed because the class is templated with the Request and Response type I guess?